atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java

```diff
@@ -0,0 +1,205 @@
+                return PtBytes.toLong(currentBackupBound);
+            }
+
+            byte[] backupValue = MoreObjects.firstNonNull(currentBound,
+                    PtBytes.toBytes(CassandraTimestampUtils.INITIAL_VALUE));
+
+            Map<String, Pair<byte[], byte[]>> casMap =
+                    ImmutableMap.of(
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            Pair.create(currentBound, CassandraTimestampUtils.INVALIDATED_VALUE.toByteArray()),
+                            CassandraTimestampUtils.BACKUP_COLUMN_NAME,
+                            Pair.create(currentBackupBound, backupValue));
+            executeAndVerifyCas(client, casMap);
+            log.info("[BACKUP] Backed up the value {}", currentBackupBound);
+            return PtBytes.toLong(backupValue);
+        });
+    }
+
+    /**
+     * Restores a backup of an existing timestamp, if possible.
+     *
+     * This method assumes that the timestamp table exists.
+     */
+    public synchronized void restoreFromBackup() {
+        clientPool().runWithRetry(client -> {
+            BoundData boundData = getCurrentBoundData(client);
+            byte[] currentBackupBound = boundData.backupBound();
+
+            BoundReadability boundReadability = checkReadability(boundData);
+            if (boundReadability == BoundReadability.BOUND) {
+                log.info("[RESTORE] Didn't restore, because the current bound is readable.");
+                return null;
+            }
+
+            Map<String, Pair<byte[], byte[]>> casMap = ImmutableMap.of(
+                    CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                    Pair.create(CassandraTimestampUtils.INVALIDATED_VALUE.toByteArray(), currentBackupBound),
+                    CassandraTimestampUtils.BACKUP_COLUMN_NAME,
+                    Pair.create(currentBackupBound, PtBytes.EMPTY_BYTE_ARRAY));
+            executeAndVerifyCas(client, casMap);
+            log.info("[RESTORE] Restored the value {}", currentBackupBound);
+            return null;
+        });
+    }
+
+    private BoundReadability checkReadability(BoundData boundData) {
+        BoundReadability boundReadability = getReadability(boundData);
+
+        Preconditions.checkState(boundReadability != BoundReadability.BOTH,
+                "We had both backup and active timestamp bounds readable! This is unexpected. Please contact support.");
+        Preconditions.checkState(boundReadability != BoundReadability.NEITHER,
+                "We had an unreadable active timestamp bound with no backup! This is unexpected. Please contact "
+                        + "support.");
+        return boundReadability;
+    }
+
+    private BoundReadability getReadability(BoundData boundData) {
+        boolean boundReadable = boundData.bound() == null
+                || CassandraTimestampUtils.isValidTimestampData(boundData.bound());
```

Comment:
Isn't it possible that this is run prior to a boundstore being run and migrating the value?  and therefore would view the value to be invalid?